src/Buffer.mo

```diff
@@ -1,166 +1,1759 @@
+        elems := Prim.Array_init<?X>(elemsSize / DOWNSIZE_FACTOR, null);
+      };
+
+      var i = 0;
+      var j = 0;
+      while (i < count) {
+        if (keep[i]) {
+          elems[j] := original[i];
+          i += 1;
+          j += 1;
+        } else {
+          i += 1;
+        }
+      };
+
+      count -= removeCount;
+    };
+
+    /// Gets the `i`-th element of this buffer. Traps if  `i >= count`. Indexing is zero-based.
+    ///
+    /// Runtime: O(1)
+    ///
+    /// Space: O(1)
+    public func get(i : Nat) : X {
+      switch (elems[i]) {
+        case (?element) element;
+        case null Prim.trap("Buffer index out of bounds in get");
+      }
+    };
+
+    /// Gets the `i`-th element of the buffer as an option. Returns `null` when `i >= count`. Indexing is zero-based.
+    ///
+    /// Runtime: O(1)
+    ///
+    /// Space: O(1)
+    public func getOpt(i : Nat) : ?X {
+      if (i < count) {
+        elems[i]
+      } else {
         null
+      }
+    };
+
+    /// Overwrites the current value of the `i`-entry of  this buffer with `elem`. Traps if the
+    /// index is out of bounds. Indexing is zero-based.
+    ///
+    /// Runtime: O(1)
+    ///
+    /// Space: O(1)
+    public func put(i : Nat, elem : X) {
+      if (i >= count) {
+        Prim.trap "Buffer index out of bounds in put";
+      };
+      elems[i] := ?elem;
+    };
+
+    /// Returns the size of the underlying array.
+    ///
+    /// Runtime: O(1)
+    ///
+    /// Space: O(1)
+    public func capacity() : Nat = elems.size();
+
+    /// Resizes the underyling array to `size`. Traps if `size` is less than
+    /// the current size of the buffer.
+    ///
+    /// Runtime: O(size)
+    ///
+    /// Space: O(size)
+    public func resize(size : Nat) {
+      if (size < count) {
+        Prim.trap "size must be >= current buffer size in resize"
+      };
+
+      let elems2 = Prim.Array_init<?X>(size, null);
+
+      var i = 0;
+      if (elems.size() <= size) {
+        label iter for (element in elems.vals()) {
+          if (i >= count) {
+            break iter;
+          };
+          elems2[i] := element;
+          i += 1;
+        }
       } else {
-        count -= 1;
-        ?elems[count]
+        while (i < count) {
+          elems2[i] := elems[i];
+          i += 1;
+        };
       };
+      elems := elems2;
     };
 
     /// Adds all elements in buffer `b` to this buffer.
-    public func append(b : Buffer<X>) {
-      let i = b.vals();
-      loop {
-        switch (i.next()) {
-          case null return;
-          case (?x) { add(x) };
+    /// 
+    /// Amortized Runtime: O(size2), Worst Case Runtime: O(size1 + size2)
+    ///
+    /// Amortized Space: O(1), Worst Case Space: O(size1 + size2)
+    public func append(buffer2 : Buffer<X>) {
+      let count2 = buffer2.size();
+      // Make sure you only resize once
+      if (count + count2 > elems.size()) {
+        // FIXME would be nice to have a tabulate for var arrays here
```

Comment:
I think we can probably implement that using the same primitive, with a different type asserted.

Reply:
Okay. I think I'll leave this out of this PR, but I'll leave the note in there for future optimization.